### ISSUE TITLE
Unable to read/display TGA as RGB with imageio-tga 3.1.1

### ISSUE THREAD

**Issue by magner58** [haraldk/TwelveMonkeys#156]
I try to read an image from a *.tga file http://www.filedropper.com/imageoriginal and display it on a JLabel
```
File file = new File("...\\image_original.tga");
BufferedImage rawImage = ImageIO.read(file);
JLabel lblNewLabel = new JLabel();
lblNewLabel.setIcon(new ImageIcon(rawImage ));
```
it is then displayed like that: http://www.filedropper.com/imagereadjlabel

When I write it to a file
```
File file_raw = new File("...\\image_read_written.jpg");
ImageIO.write(rawImage, "JPG", file_raw);
```
the result is: http://www.filedropper.com/imagereadwritten

I tried to follow your advice from https://github.com/haraldk/TwelveMonkeys/issues/155 and convert the image, but the result is blackish for all image types except for BufferedImage.TYPE_BYTE_GRAY:
```
File file = new File("...\\image_original.tga");
BufferedImage rawImage = ImageIO.read(file);
JLabel lblNewLabel = new JLabel();
BufferedImage convertedImage = new BufferedImage(rawImage.getWidth(), rawImage.getHeight(), BufferedImage.TYPE_BYTE_GRAY);
ColorConvertOp colorConvert = new ColorConvertOp(null);
colorConvert.filter(rawImage, convertedImage);
lblNewLabel.setIcon(new ImageIcon(convertedImage));
```
displayed like:http://www.filedropper.com/imagereadconvertedjlabel

**Comment by haraldk** [comment]
Hi,

Thanks for reporting!

Seems your TGA file has the alpha channel stored inversed, compared to all samples I've got... Do you know what software created it? Does it open properly in other software?

I could easily apply a fix that displays your TGA correctly, however, that would break a lot of my existing samples... So I need to read up on the specs to figure out what to do.


Regards,

-- 
Harald K

**Comment by magner58** [comment]
Hey, also thought it has something to do with alpha :) Not sure how it was created - that's how all images in "unpacked" (for modding) Medieval 2 Total War are stored. Could be that they did something wrong, but at least my standard viewer (ACDSee 5.0) displays the original image (like that http://www.filedropper.com/imageoriginal_1) correctly by default.

P.S. Is there a workaround I could use to reverse the alpha channel before displaying it?

**Comment by haraldk** [comment]
I see. 

Well, you could.. Reverse the alpha channel. :-)

Or, you could go to `TGAImageReader` line 243 (the `invertAlpha` method), and just make it a no-op (remove/comment out the method body):

    private void invertAlpha(final byte[] rowDataByte) {
    //    for (int i = 3; i < rowDataByte.length; i += 4) {
    //        rowDataByte[i] = (byte) (0xFF - rowDataByte[i]);
    //    }
    }

This way your images will look right when displaying. Also, they will look okay when saved as PNG, if you want to store as JPEG, you should remove the alpha channel altogether. 

PS: You can attach images in the issue tracker. It's easier for you, and I won't have to enter the stupid FileDropper captcha for *each and every* download. Just saying... 

-- 
Harald K

**Comment by haraldk** [comment]
Okay, after consulting the specs, I have agreed with myself that this is a bug, and that the alpha should only be changed when the "attribute bits" of the header is `0` (your image has a value of `8`, meaning there's a full 8 bit alpha channel present).

For some of my test images, this makes them all transparent, but I think these images are actually wrong (even though the meta data says `"Truevision(R) Sample Image"`). 

-- 
Harald K

**Comment by magner58** [comment]
> go to TGAImageReader line 243

Thanks, will give it a try :)

> you could reverse the alpha channel

Is it possible to do it on the BufferedImage rawImage = ImageIO.read(file) with something like ColorConvertOp? Sorry for stupid questions, I am not that good with graphics :)

> this is a bug, and that the alpha should only be changed when the attribute bits of the header is 0 ... these images are actually wrong 

I probably could kinda confirm that - my ACDSee can't display any of TGA Images from your tests (like XING_B24.TGA) and it also does not show file type info.

P.S. I tried to attach original image, but gitbub tells me "Unfortunately, we don’t support that file type. Try again with a PNG, GIF, JPG" :( But it should work with non-tga, you are right, I simpy didn't think about it.

**Comment by haraldk** [comment]
> Is it possible to do it on the BufferedImage rawImage = ImageIO.read(file) with something like ColorConvertOp? 

Sure, but that would be re-reversing the already reversed alpha. :-)

As I know the 32 bit TGA images read with my plugin are using a `DataBufferByte` backing buffer, you could do this (but it will most likely break in the future, when I fix the bug, or if the data type changes):

    BufferedImage image = ImageIO.read(file);

    // Get pixels as interleaved BGRA byte array (assumes TGA read by my reader only)
    byte[] pixels = ((DataBufferByte) image.getRaster().getDataBuffer()).getData();

    // Reverse the last component of each pixel (255 - val)
    for (int i = 3; i < pixels.length; i += 4) {
        pixels[i] = (byte) (0xFF - pixels[i]);
    }

The sample images I use, are from http://www.fileformat.info/format/tga/sample/index.htm, they should be mostly good.

> I tried to attach original image, but gitbub tells me "Unfortunately, we don’t support that file type.

Yes, I know the file support in the issue tracker is quite lame... But it's easy to work around. Just rename any non-PNG/GIF/JPEG file like `file.ext` to something like `file_ext.png`, and it will attach the file as an image (won't look good in the browser, but it works). :-)

-- 
Harald K

-- 
Harald K

**Comment by haraldk** [comment]
Fixed in fd4745f6a663bcb197c01039ac8013dd8618eb5c.

-- 
Harald K

**Comment by magner58** [comment]
Great, thanks for the fix and workaround!

Workaround works nicely, will put a TODO and wait for the next twelvemonkeys imageio release :)

P.S. Btw image is now better displayed in my JLaben than in ACDsee - it has better transparent box:
![image](https://cloud.githubusercontent.com/assets/13498110/8925661/c554b9fc-3509-11e5-946f-3c96af360f1e.jpg)
